VM guidance per capability - first release (#533)
microsoft/finops-toolkit: docs/services/compute/virtual-machines.md

- **Backup:** Backup costs are associated with the chosen backup replication policy and frequency. Regularly assess and adjust backup strategies to align with cost-effectiveness and data recovery needs.

### Allocation

<!---
Add link to MS or F2 docs explaining the capability
-->

A highly effective method for grouping and allocating costs is through the tag inheritance feature in Cost Management. Enabled tag inheritance applies billing, resource group, and subscription tags to child resource usage records.

Once tag inheritance is enabled, tags from billing profiles, invoice sections, subscriptions, and resource groups are automatically applied to child resource usage records for the current month within 24 hours. For detailed instructions, refer to the [Cost allocation guidance](https://learn.microsoft.com/en-us/azure/cost-management-billing/costs/enable-tag-inheritance).

#### Tagging best practices:

- **Consistent tagging:** Establish a consistent tagging strategy across VMs using the [Cloud Adoption Framework (CAF) guidance](https://learn.microsoft.com/en-us/azure/cloud-adoption-framework/ready/azure-best-practices/resource-tagging).

> Implementing tag inheritance and following these tagging best practices enhances your ability to efficiently allocate and manage VM costs.

Reviewer: You can get rid of all this since it's not specific to VMs.

### Reporting + analytics

<!---

> Side note: Cost analysis has a feature where you can add a "cm-resource-parent" tag to your child resources (e.g., storage accounts) to link them to their logical parent and then Cost analysis will show the costs grouped together. We should inform them about this somewhere. It should be a separate doc that we reference probably in the reporting capability and possibly from here, if we end up calling out the issue.

Reviewer: ```suggestion
TODO: Cost analysis has a feature where you can add a "cm-resource-parent" tag to your child resources (e.g., storage accounts) to link them to their logical parent and then Cost analysis will show the costs grouped together. We should inform them about this somewhere. It should be a separate doc that we reference probably in the reporting capability and possibly from here, if we end up calling out the issue.
```

-->

To report on VM costs comprehensively, consider all the services mentioned in the Data Ingestion capability, including but not limited to:

- **Virtual machines:** Understand the hourly billing, VM size, regional pricing, and states (running, stopped, deallocated).
- **Storage:** Factor in storage costs based on disk size and transactions per hour. 
- **Virtual network and bandwidth:** Include costs related to data transfer and bandwidth usage.
- **Associated services:** Consider costs associated with Azure Monitor, Azure Bastion, Azure DNS, and Load Balancer if used.

## Validating VM usage:

Once you've identified all the costs associated with a VM, the next crucial step is to validate the VM usage to ensure it is properly sized. This involves:

- **Right-sizing VMs:** Review the VM size to ensure it aligns with the actual workload requirements. Consider scaling up or down based on usage patterns.
- **Efficient resource management:** Leverage Azure services such as Cost Management and Azure Monitor to gain insights into VM performance metrics, allowing you to identify areas for optimization.

### Anomalies
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->
Anomaly detection helps pinpoint unexpected spikes or irregularities in VM spending, enabling timely intervention to optimize costs. Identifying irregularities helps in optimizing resource allocation, ensuring resources are used efficiently.

> 1. **Utilize Azure monitor:**
>    - Leverage Azure monitor to access detailed insights into VM performance metrics. Set up alerts for key metrics such as CPU usage, memory utilization, and disk activity.

Reviewer: Are there useful links, API calls, or queries we could add here? Especially if they can do real anomaly detection but simple delta/variance also works. This could be a nice template to offer 🙂
```suggestion
1. **Utilize Azure Monitor:**
   - Leverage Azure Monitor to access detailed insights into VM performance metrics. Set up alerts for key metrics such as CPU usage, memory utilization, and disk activity.
```

> 2. **Cost management tools:**

Reviewer: We can probably remove this since there isn't anything specific to VMs here.

   - Leverage the cost management tools, such as Azure Cost Management and Power BI, to analyze spending patterns and detect unusual spikes in VM costs.

3. **Performance metrics:**
   - Correlate cost anomalies with performance metrics. Anomalies in resource utilization may directly impact costs.

4. **Resource configuration:**

>    - Evaluate the configuration of VM resources. Anomalies could be linked to improper resource allocation or sudden changes in demand.

Reviewer: Would be good to get links, queries, APIs added for these as well.

<br>

## Quantify business value

### Unit economics

<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### Estimation + forecasting
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

Before creating a VM, leverage the [Azure Pricing Calculator](https://azure.microsoft.com/en-us/pricing/calculator/) to simulate costs. This tool allows you to estimate expenses based on your selected VM configurations, helping you make informed decisions and plan budgets effectively.

Understanding and proactively managing the costs associated with VMs and related services are key components of a successful cloud financial management strategy. Regularly review and optimize your resource usage to ensure cost-effectiveness and alignment with your business goals.

### Budgeting
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

> Effectively budgeting for VMs in Azure involves considersing different factors, including VM sizes, usage patterns, and potential discounts.

Reviewer: This doesn't add much value. Should we remove it?

### Benchmarking
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->
<br>

## Quantify business value
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### Architecting for cloud

<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

> ## Understanding VM sizes (SKUs):

Reviewer: Do these headers add any value? Could we just keep this as a simple bulleted list?

Side note: Don't add an H2 under an H3 😜

- **Evaluate performance needs:** Choose VM sizes based on the specific performance requirements of your workloads. Consider factors such as CPU, memory, and storage capabilities.

## Usage patterns:

- **Continuous vs. scheduled usage:**
   - Determine if VMs need to run continuously or if there are periods when they can be powered off. Schedule VMs to align with actual workload demands, optimizing costs during idle periods.

> ## Existing commitment-based discounts:

Reviewer: They're getting rid of "-based"
```suggestion
## Existing commitment discounts:
```

- **Reserved instances (RIs):**

>    - Leverage Reserved Instances for VMs to benefit from significant cost savings. Commit to a one or three-year term for VMs that require ongoing, predictable usage.

Reviewer: ```suggestion
   - Leverage reservations for VMs that have consistent usage patterns to benefit from significant cost savings. Commit to a one-, three-, or five-year term for VMs with ongoing, predictable usage.
```

> - **Azure hybrid benefit (AHB):**

Reviewer: ```suggestion
- **Azure Hybrid Benefit (AHB):**
```

   - If you have on-premises Windows Server licenses with Software Assurance, consider utilizing Azure Hybrid Benefit for additional cost savings on VMs.

## Underlying services:

- **Networking costs:**
   - Factor in networking costs associated with data transfer and bandwidth. Optimize network configurations to minimize unnecessary expenses.

- **Backup services:**
   - Assess backup requirements for VMs. Utilize Azure Backup to efficiently manage and optimize backup costs based on retention policies and storage needs.

- **Monitoring and alerts:**
   - Leverage Azure Monitor for VMs to gain insights into performance metrics and set up alerts. Proactively monitor VMs to identify potential issues and optimize resource usage.

### Utilization efficiency
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

> Optimizing the utilization efficiency of VMs in Azure is essential for achieving cost savings and maximizing resource efficiency. 

Reviewer: Not needed

1. **Regular right-sizing reviews:**
   - Conduct regular reviews of VM sizes based on performance assessments and Azure Advisor recommendations. Right-size VMs to align with actual usage patterns.

<!---

> Might also be good to call out what recommendations are available in Advisor and include some ARG queries for pulling them.

Reviewer: ```suggestion
TODO: Might also be good to call out what recommendations are available in Advisor and include some ARG queries for pulling them.
```

-->

2. **Scheduled shutdowns:**
   - Identify and schedule automated VM shutdowns during idle periods. Implement Azure Automation runbooks for efficient management of startup and shutdown routines.

3. **Continuous monitoring and alerts:**
   - Implement continuous monitoring using Azure Monitor and set up

> 4. - **Automated tagging with Azure policy:** Simplify and standardize tagging by utilizing Azure Policy to automatically apply tags to VMs. This streamlines the allocation process, ensuring adherence to tagging standards.

Reviewer: I'm assuming you meant to keep this on 2 lines like the others...
```suggestion
4. **Automated tagging with Azure policy:**
   - Simplify and standardize tagging by utilizing Azure Policy to automatically apply tags to VMs. This streamlines the allocation process, ensuring adherence to tagging standards.
```

### Workload management

<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

<!---

>  add new feature to a tool: something in the workload management section about moving VMs between regions based on regional pricing. That's probably a good opportunity for a workbook, PBI report, or AOE recommendation :-)

Reviewer: ```suggestion
TODO: add new feature to a tool: something in the workload management section about moving VMs between regions based on regional pricing. That's probably a good opportunity for a workbook, PBI report, or AOE recommendation :-)
```

>  There's also an opportunity for us to build some tooling to recommend cm-resource-parent tags you should apply to your resources, then offering an approval workflow to apply them on your behalf.

Reviewer: ```suggestion
TODO: There's also an opportunity for us to build some tooling to recommend cm-resource-parent tags you should apply to your resources, then offering an approval workflow to apply them on your behalf.
```

-->

1. **Regular performance reviews:**
   - Conduct regular reviews of VM performance metrics. Adjust VM SKU based on changing workload patterns.

2. **Automation implementation:**
<!---

> - FYI - Autostop is part of the workload management capability. Just calling it out in case I forget by the time I get down to that section :-)

Reviewer: ```suggestion
TODO: Move to workload management
```

-->

   - Implement automation for routine workload management tasks. This includes auto stop and start VMs based on schedule.

3. **Continuous monitoring:**
   - Establish a continuous monitoring process using Azure Monitor and Log Analytics. Proactively address performance issues to maintain optimal workload management.

### Licensing + SaaS

<!-- TODO: Add AHB and update based on F2 licensing guidance-->

### Commitment discounts
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

<!---

> Let's get more specific and actionable. ARG queries, PBI reports, etc.

Reviewer: ```suggestion
TODO: Let's get more specific and actionable. ARG queries, PBI reports, etc.
```

-->

> Commitment-based discounts, such as Reserved Instances (RIs) and Azure Savings Plans, offer significant cost savings for organizations utilizing VMs in Azure.

Reviewer: nit: Technically, it's "reservation" and not "reserved instance". Using "savings plan" in lowercase for consistency.
```suggestion
Commitment discounts, such as reservations and savings plans, offer significant cost savings for organizations utilizing VMs in Azure.
```

1. **Evaluate workload stability:**
   - Assess VM workloads to understand stability and usage patterns. Stable workloads with predictable usage are prime candidates for Reserved Instances.

> 2. **Long-Term commitments:**

Reviewer: ```suggestion
2. **Long-term commitments:**
```

   - If your organization can commit to long-term usage, consider Reserved Instances. Evaluate the one or three-year commitment options based on budgetary considerations.

3. **Flexibility and usage variation:**
   - For workloads with varying usage or the need for flexibility, explore Azure Savings Plans. Savings Plans provide discounts while accommodating changes in VM size, family, or region.

4. **Cost analysis:**

>    - Utilize tools like Azure Pricing Calculator and Azure Cost Management to perform cost analysis. Compare the potential savings of RIs and Savings Plans against pay-as-you-go pricing.

Reviewer: ```suggestion
   - Utilize tools like Azure Pricing Calculator and Microsoft Cost Management to perform cost analysis. Compare the potential savings of commitment discounts against on-demand pricing.
```

5. **Regular review:**

>    - Periodically review VM workloads and commitments. Adjust Reserved Instances or Savings Plans as needed to align with evolving usage patterns.

Reviewer: ```suggestion
   - Periodically review VM workloads and commitments. Adjust commitment discounts as needed to align with evolving usage patterns.
```

### Cloud sustainability
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

<br>

## Manage the FinOps practice
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### FinOps practice operations
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### FinOps education + enablement
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### FinOps assessment
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### Cloud policy + governance
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### FinOps tools + services
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### Chargeback + invoicing
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

-->

### Intersecting disciplines
<!---

> Add link to MS or F2 doc explaining the capability

Reviewer: ```suggestion
TODO: Add link to MS or F2 doc explaining the capability
```

## Understand cloud usage and cost

### Data ingestion

> **Understanding and Managing Costs for Virtual Machines**

Author: Option 3
Move it to be body text under ## Understand, but not in an explicit subsection.

Azure virtual machines (VMs) are a fundamental component of your cloud infrastructure, providing on-demand computing resources tailored to your specific needs. To effectively manage costs associated with VMs, it's crucial to understand the various factors that contribute to their pricing, as well as costs related to associated services.

## Virtual Machines:

- **Hourly Billing:** VMs are billed based on the number of hours they are provisioned. The longer a VM runs, the higher the associated cost.

- **VM Size:** The pricing of VMs varies based on the chosen size or configuration. Larger and more powerful VMs generally incur higher costs.

- **Regional Pricing:** The cost of a VM is influenced by the region where it is deployed. Different regions may have varying pricing structures, impacting the overall cost of VM usage.

- **VM States:** As VM instances transition through different states (running, stopped, deallocated), they may incur different billing rates. Ensure efficient VM management to minimize unnecessary costs during idle periods.

> ## Storage:

Author: Done!

- **Disk Size:** Storage costs are determined by the size of disks attached to VMs, measured in gigabytes (GB). Larger disk sizes contribute to higher storage costs.

- **Transactions:** Transaction-based costs apply to storage and are influenced by the number of transactions per hour. Refer to the [storage billing documentation](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-billing) for detailed information on transaction costs.

> ## Virtual Network and Bandwidth:

Author: Yes, they are handled differently. Let me know if it is clearer now.

- **Bandwidth:** Similar to virtual network data transfer, bandwidth costs are based on the amount of data transferred. Efficient network management helps control bandwidth-related expenses.

## Associated Services:

- **Azure Monitor:** Costs for Azure Monitor are based on the volume of data ingested. Monitor and optimize the data ingested to manage associated expenses effectively.

- **Azure Bastion:** Billing for Azure Bastion is determined by the volume of data transferred. Monitor and control data transfer to regulate costs.

- **Azure DNS:** Charges for Azure DNS depend on the number of DNS zones hosted and the quantity of DNS queries received. Manage DNS resources efficiently to control costs.

- **Load Balancer:** If load balancers are utilized, costs are determined by the number of rulesets, hours used, and the amount of data processed. Optimize load balancer configurations for cost-effectiveness.

- **Backup:** Backup costs are associated with the chosen backup replication policy and frequency. Regularly assess and adjust backup strategies to align with cost-effectiveness and data recovery needs.

### Allocation

> Efficient cost allocation is crucial for managing Azure Virtual Machine (VM) expenses. A highly effective method for grouping and allocating costs is through the tag inheritance feature in Cost Management. Enabled tag inheritance applies billing, resource group, and subscription tags to child resource usage records.

Author: Maybe we could replace these statements with a link to the capability doc (either MS doc or F2 doc). I will replace them with placeholders and we can discuss it later

To report on VM costs comprehensively, consider all the services mentioned in the Data Ingestion capability, including but not limited to:

- **Virtual Machines:** Understand the hourly billing, VM size, regional pricing, and states (running, stopped, deallocated).
- **Storage:** Factor in storage costs based on disk size and transactions per hour.
- **Virtual Network and Bandwidth:** Include costs related to data transfer and bandwidth usage.
- **Associated Services:** Consider costs associated with Azure Monitor, Azure Bastion, Azure DNS, and Load Balancer if used.

## Validating VM Usage:

Once you've identified all the costs associated with a VM, the next crucial step is to validate the VM usage to ensure it is properly sized. This involves:

- **Right-sizing VMs:** Review the VM size to ensure it aligns with the actual workload requirements. Consider scaling up or down based on usage patterns.
- **Efficient Resource Management:** Leverage Azure services such as Cost Management and Azure Monitor to gain insights into VM performance metrics, allowing you to identify areas for optimization.

### Anomalies

> Anomaly detection helps pinpoint unexpected spikes or irregularities in VM spending, enabling timely intervention to optimize costs. IIdentifying irregularities helps in optimizing resource allocation, ensuring resources are used efficiently.

Reviewer: This is all generic. We can probably remove it like I mentioned for allocation above. Maybe every section would have a link to the general FinOps capability guidance 🤔

Author: Added placeholders to include the links to the FinOps capability guidance. Will do that in a later PR.

1. **Utilize Azure Monitor:**
   - Leverage Azure Monitor to access detailed insights into VM performance metrics. Set up alerts for key metrics such as CPU usage, memory utilization, and disk activity.

2. **Cost Management Tools:**
   - Leverage the cost management tools, such as Azure Cost Management and Power BI, to analyze spending patterns and detect unusual spikes in VM costs.

3. **Performance Metrics:**
   - Correlate cost anomalies with performance metrics. Anomalies in resource utilization may directly impact costs.

4. **Resource Configuration:**
   - Evaluate the configuration of VM resources. Anomalies could be linked to improper resource allocation or sudden changes in demand.

<br>

> ## Optimize cloud usage and cost

Reviewer: My bad. I accidentally swapped these. Can you also change the one below? I can't leave a suggestion on that line.
```suggestion
## Quantify business value
```

> ### Unit economics

Reviewer: nit: We should probably move unit economics to be last in this section since it's the most complex topic. I like the idea of having them sorted by logical progression.

### Estimation + forecasting

> Before creating a VM, leverage the [Azure Pricing Calculator](https://azure.microsoft.com/en-us/pricing/calculator/) to simulate costs. This tool allows you to estimate expenses based on your selected VM configurations, helping you make informed decisions and plan budgets effectively.

Reviewer: Might be better to link to the pricing pages for each of the different resource types here. APC is covered in the general guidance already.

Understanding and proactively managing the costs associated with virtual machines and related services are key components of a successful cloud financial management strategy. Regularly review and optimize your resource usage to ensure cost-effectiveness and alignment with your business goals.

### Budgeting

> Effectively budgeting for Virtual Machines (VMs) in Azure involves considersing different factors, including VM sizes, usage patterns, and potential discounts.

Reviewer: nit: This intro isn't needed. The rest of the content should move to Architecting for cloud.

### Utilization efficiency

Optimizing the utilization efficiency of Virtual Machines (VMs) in Azure is essential for achieving cost savings and maximizing resource efficiency. 

> 1. **Commitment-Based Discounts:**

Reviewer: This should only be in the Commitment discounts section below.

Author: removed.

   - Evaluate workloads for suitability with Reserved Instances or Spot Instances. Leverage commitment-based discounts to optimize costs for consistent and variable workloads.

2. **Regular Right-Sizing Reviews:**

>    - Conduct regular reviews of VM sizes based on performance assessments and Azure Advisor recommendations. Right-size VMs to align with actual usage patterns.

Reviewer: Might also be good to call out what recommendations are available in Advisor and include some ARG queries for pulling them.

Author: added a placeholder for that.

> 3. **Scheduled Shutdowns:**

Reviewer: Move to workload management

Author: moved.

   - Identify and schedule automated VM shutdowns during idle periods. Implement Azure Automation runbooks for efficient management of startup and shutdown routines.

> 4. **Continuous Monitoring and Alerts:**

Reviewer: Conceptually, this is good, but it's not specific to VMs. We should probably be specific or just point to general guidance.

   - Implement continuous monitoring using Azure Monitor and set up

### Workload management

> 1. **Regular Performance Reviews:**

Reviewer: Let's replace the generic guidance with something more targeted for VMs. Might be good to call out auto-stop and auto-scale features for VMs specifically.

   - Conduct regular reviews of VM performance metrics. Adjust VM SKU based on changing workload patterns.

2. **Automation Implementation:**
   - Implement automation for routine workload management tasks. This includes auto stop and start VMs based on schedule.

3. **Continuous Monitoring:**
   - Establish a continuous monitoring process using Azure Monitor and Log Analytics. Proactively address performance issues to maintain optimal workload management.

> ### Licensing + SaaS

Reviewer: Need AHB (can be in a future PR). Maybe at least add a TODO comment. Also calling out some licensing guidance F2 is adding that might be useful here. I can share that later.
```suggestion
### Licensing + SaaS

<!-- TODO: Add AHB and update based on F2 licensing guidance-->

```

### Commitment discounts

> Commitment-based discounts, such as Reserved Instances (RIs) and Azure Savings Plans, offer significant cost savings for organizations utilizing Virtual Machines (VMs) in Azure.

Reviewer: Let's get more specific and actionable. ARG queries, PBI reports, etc.

1. **Evaluate Workload Stability:**
   - Assess VM workloads to understand stability and usage patterns. Stable workloads with predictable usage are prime candidates for Reserved Instances.

2. **Long-Term Commitments:**
   - If your organization can commit to long-term usage, consider Reserved Instances. Evaluate the one or three-year commitment options based on budgetary considerations.

3. **Flexibility and Usage Variation:**
   - For workloads with varying usage or the need for flexibility, explore Azure Savings Plans. Savings Plans provide discounts while accommodating changes in VM size, family, or region.

4. **Cost Analysis:**
   - Utilize tools like Azure Pricing Calculator and Azure Cost Management to perform cost analysis. Compare the potential savings of RIs and Savings Plans against pay-as-you-go pricing.

5. **Regular Review:**
   - Periodically review VM workloads and commitments. Adjust Reserved Instances or Savings Plans as needed to align with evolving usage patterns.

### Cloud sustainability

> <br>

Reviewer: For the rest...
1. Do you know if the assessment had anything specific to VMs? If so, might be worth pointing to it.
2. I'm sure policy has some built-in options that might be useful here. If not, that's definitely an idea for the toolkit.
3. Do we want to reference third-party tools/services that help with VM cost management? There are a lot and we won't be all-inclusive. We could start with the top 5 in the Marketplace and let people submit PRs. Or we just avoid it altogether 😜
4. Comment out any sections that don't have content.
